mingw-w64: Many `run-make-fulldeps` tests fail due to some change in library search

This is a regression from the last 1-2 weeks.
Not sure who is the right person to ping, @alexcrichton @michaelwoerister maybe.

Failing tests:
```
[run-make] run-make-fulldeps\a-b-a-linker-guard
[run-make] run-make-fulldeps\c-dynamic-dylib
[run-make] run-make-fulldeps\c-static-dylib
[run-make] run-make-fulldeps\dylib-chain
[run-make] run-make-fulldeps\extern-fn-reachable
[run-make] run-make-fulldeps\issue-15460
[run-make] run-make-fulldeps\issue-19371
[run-make] run-make-fulldeps\mixing-deps
[run-make] run-make-fulldeps\prefer-dylib
[run-make] run-make-fulldeps\simple-dylib
[run-make] run-make-fulldeps\static-nobundle
```
Typical failure looks like
```
---- [run-make] run-make-fulldeps\c-dynamic-dylib stdout ----

error: make failed
status: exit code: 2
command: "make"
stdout:
------------------------------------------
gcc.exe -ffunction-sections -fdata-sections -m64 -c -o /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/libcfoo.o cfoo.c
gcc.exe -ffunction-sections -fdata-sections -m64 -o /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/cfoo.dll /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/libcfoo.o -shared
PATH="/home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu:C:/msys64/home/we/rust/build\x86_64-pc-windows-gnu\stage2\bin:/home/we/rust/build/x86_64-pc-windows-gnu/stage0-tools/x86_64-pc-windows-gnu/release/deps:/home/we/rust/build/x86_64-pc-windows-gnu/stage0-sysroot/lib/rustlib/x86_64-pc-windows-gnu/lib:/mingw64/bin:/c/Users/we/.cargo/bin:/c/Program Files/CMake/bin:/c/Program Files/Git/cmd:/c/mingw-w64/x86_64-7.2.0-posix-seh-rt_v5-rev1/mingw64/bin:/c/ProgramData/chocolatey/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/usr/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" 'C:/msys64/home/we/rust/build\x86_64-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu -L /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu  foo.rs -C prefer-dynamic
PATH="/home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu:C:/msys64/home/we/rust/build\x86_64-pc-windows-gnu\stage2\bin:/home/we/rust/build/x86_64-pc-windows-gnu/stage0-tools/x86_64-pc-windows-gnu/release/deps:/home/we/rust/build/x86_64-pc-windows-gnu/stage0-sysroot/lib/rustlib/x86_64-pc-windows-gnu/lib:/mingw64/bin:/c/Users/we/.cargo/bin:/c/Program Files/CMake/bin:/c/Program Files/Git/cmd:/c/mingw-w64/x86_64-7.2.0-posix-seh-rt_v5-rev1/mingw64/bin:/c/ProgramData/chocolatey/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/usr/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" 'C:/msys64/home/we/rust/build\x86_64-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu -L /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu  bar.rs
PATH="/home/we/rust/build/x86_64-pc-windows-gnu/stage0-tools/x86_64-pc-windows-gnu/release/deps:/home/we/rust/build/x86_64-pc-windows-gnu/stage0-sysroot/lib/rustlib/x86_64-pc-windows-gnu/lib:/mingw64/bin:/c/Users/we/.cargo/bin:/c/Program Files/CMake/bin:/c/Program Files/Git/cmd:/c/mingw-w64/x86_64-7.2.0-posix-seh-rt_v5-rev1/mingw64/bin:/c/ProgramData/chocolatey/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/usr/bin:/c/Windows/System32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:C:/msys64/home/we/rust/build\x86_64-pc-windows-gnu\stage2\lib\rustlib\x86_64-pc-windows-gnu\lib" /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/bar
rm /home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/libcfoo.o

------------------------------------------
stderr:
------------------------------------------
C:/msys64/home/we/rust/build/x86_64-pc-windows-gnu/test/run-make-fulldeps/c-dynamic-dylib.stage2-x86_64-pc-windows-gnu/bar.exe: error while loading shared libraries: foo.dll: cannot open shared object file: No such file or directory
make: *** [Makefile:11: all] Error 127

------------------------------------------
```

--- a/src/bootstrap/config.rs
+++ b/src/bootstrap/config.rs
@@ -325,6 +325,14 @@ struct TomlTarget {
 }
 
 impl Config {
+    fn path_from_python(var_key: &str) -> PathBuf {
+        match env::var_os(var_key) {
+            // Do not trust paths from Python and normalize them slightly (#49785).
+            Some(var_val) => Path::new(&var_val).components().collect(),
+            _ => panic!("expected '{}' to be set", var_key),
+        }
+    }
+
     pub fn default_opts() -> Config {
         let mut config = Config::default();
         config.llvm_enabled = true;
@@ -348,9 +356,9 @@ impl Config {
         config.deny_warnings = true;
 
         // set by bootstrap.py
-        config.src = env::var_os("SRC").map(PathBuf::from).expect("'SRC' to be set");
         config.build = INTERNER.intern_str(&env::var("BUILD").expect("'BUILD' to be set"));
-        config.out = env::var_os("BUILD_DIR").map(PathBuf::from).expect("'BUILD_DIR' set");
+        config.src = Config::path_from_python("SRC");
+        config.out = Config::path_from_python("BUILD_DIR");
 
         let stage0_root = config.out.join(&config.build).join("stage0/bin");
         config.initial_rustc = stage0_root.join(exe("rustc", &config.build));

